lib/zlog.c

```diff
@@ -0,0 +1,702 @@
+#elif defined(linux)
+	rv = syscall(__NR_gettid);
+#elif defined(__NetBSD__)
+	rv = _lwp_self();
+#elif defined(__FreeBSD__)
+	thr_self(&rv);
+#elif defined(__DragonFly__)
+	rv = lwp_gettid();
+#elif defined(__OpenBSD__)
+	rv = getthrid();
+#elif defined(__sun)
+	rv = pthread_self();
+#elif defined(__APPLE__)
+	rv = mach_thread_self();
+	mach_port_deallocate(mach_task_self(), rv);
+#endif
+	return rv;
+}
+
+void zlog_tls_buffer_init(void)
+{
+	struct zlog_tls *zlog_tls;
+	char mmpath[MAXPATHLEN];
+	int mmfd;
+	size_t i;
+
+	zlog_tls = zlog_tls_get();
+
+	if (zlog_tls || zlog_tmpdirfd < 0)
+		return;
+
+	zlog_tls = XCALLOC(MTYPE_LOG_TLSBUF, sizeof(*zlog_tls));
+	for (i = 0; i < array_size(zlog_tls->msgp); i++)
+		zlog_tls->msgp[i] = &zlog_tls->msgs[i];
+
+	snprintfrr(mmpath, sizeof(mmpath), "logbuf.%ld", zlog_gettid());
+
+	mmfd = openat(zlog_tmpdirfd, mmpath,
+		      O_RDWR | O_CREAT | O_EXCL | O_CLOEXEC, 0600);
+	fchown(mmfd, zlog_uid, zlog_gid);
+
+	if (mmfd < 0) {
+		zlog_err("failed to open thread log buffer \"%s\": %s",
+			 mmpath, strerror(errno));
+		goto out_anon;
+	}
+
+#ifdef HAVE_POSIX_FALLOCATE
+	if (posix_fallocate(mmfd, 0, TLS_LOG_BUF_SIZE) < 0) {
+#else
+	if (ftruncate(mmfd, TLS_LOG_BUF_SIZE) < 0) {
+#endif
+		zlog_err("failed to allocate thread log buffer \"%s\": %s",
+			 mmpath, strerror(errno));
+		goto out_anon_unlink;
+	}
+
+	zlog_tls->mmbuf = mmap(NULL, TLS_LOG_BUF_SIZE, PROT_READ | PROT_WRITE,
+			      MAP_SHARED, mmfd, 0);
+	if (zlog_tls->mmbuf == MAP_FAILED) {
+		zlog_err("failed to mmap thread log buffer \"%s\": %s",
+			 mmpath, strerror(errno));
+		goto out_anon_unlink;
+	}
+
+	close(mmfd);
+	zlog_tls_set(zlog_tls);
+	return;
+
+out_anon_unlink:
+	unlink(mmpath);
+	close(mmfd);
+out_anon:
+
+#ifndef MAP_ANONYMOUS
+#define MAP_ANONYMOUS MAP_ANON
+#endif
+	zlog_tls->mmbuf = mmap(NULL, TLS_LOG_BUF_SIZE, PROT_READ | PROT_WRITE,
+			      MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+
+	if (!zlog_tls->mmbuf) {
+		zlog_err("failed to anonymous-mmap thread log buffer: %s",
+			 strerror(errno));
+		XFREE(MTYPE_LOG_TLSBUF, zlog_tls);
+		zlog_tls_set(NULL);
+		return;
+	}
+
+	zlog_tls_set(zlog_tls);
+}
+
+void zlog_tls_buffer_fini(void)
+{
+	char mmpath[MAXPATHLEN];
+
+	zlog_tls_buffer_flush();
+
+	zlog_tls_free(zlog_tls_get());
+	zlog_tls_set(NULL);
+
+	snprintfrr(mmpath, sizeof(mmpath), "logbuf.%ld", zlog_gettid());
+	if (unlinkat(zlog_tmpdirfd, mmpath, 0))
+		zlog_err("unlink logbuf: %s (%d)", strerror(errno), errno);
+}
+
+#else /* !CAN_DO_TLS */
+void zlog_tls_buffer_init(void)
+{
+}
+
+void zlog_tls_buffer_fini(void)
+{
+}
+#endif
+
+static inline void zlog_tls_free(void *arg)
+{
+	struct zlog_tls *zlog_tls = arg;
+
+	if (!zlog_tls)
+		return;
+
+	munmap(zlog_tls->mmbuf, TLS_LOG_BUF_SIZE);
+	XFREE(MTYPE_LOG_TLSBUF, zlog_tls);
+}
+
+void zlog_tls_buffer_flush(void)
+{
+	struct zlog_target *zt;
+	struct zlog_tls *zlog_tls = zlog_tls_get();
+
+	if (!zlog_tls)
+		return;
+	if (!zlog_tls->nmsgs)
+		return;
+
+	rcu_read_lock();
```

Comment:
`rcu_read_lock()` is a very strange member of the "lock" family: it never blocks, the worst-case cost is a non-blocking syscall.  That's why it's lock-free: it can't be made to block by other threads.

If you haven't looked at RCU before, `rcu_read_lock()` is basically the R side of a RW lock, except the W side is replaced by a ordered / barrier list (`rcu_enqueue()`) and does not block the R side.

So, I don't think we need a log file per thread, locking is already avoided :)